airflow/contrib/hooks/sagemaker_hook.py

```diff
@@ -16,299 +16,793 @@
+            until the job completes
+        :type wait_for_completion: bool
+        :param check_interval: The interval in seconds between polling for new log entries and job completion
+        :type check_interval: int
+        :param max_ingestion_time: the maximum ingestion time in seconds. Any
+            SageMaker jobs that run longer than this will fail. Setting this to
+            None implies no timeout for any SageMaker job.
+        :type max_ingestion_time: int
+        :return: None
+        """
+
+        sec = 0
+        description = self.describe_training_job(job_name)
+        self.log.info(secondary_training_status_message(description, None))
+        instance_count = description['ResourceConfig']['InstanceCount']
+        status = description['TrainingJobStatus']
+
+        stream_names = []  # The list of log streams
+        positions = {}     # The current position in each stream, map of stream name -> position
+
+        job_already_completed = False if status in non_terminal_states else True
+
+        state = LogState.TAILING if wait_for_completion and not job_already_completed else LogState.COMPLETE
+
+        # The loop below implements a state machine that alternates between checking the job status and
+        # reading whatever is available in the logs at this point. Note, that if we were called with
+        # wait_for_completion == False, we never check the job status.
+        #
+        # If wait_for_completion == TRUE and job is not completed, the initial state is TAILING
+        # If wait_for_completion == FALSE, the initial state is COMPLETE
+        # (doesn't matter if the job really is complete).
+        #
+        # The state table:
+        #
+        # STATE               ACTIONS                        CONDITION             NEW STATE
+        # ----------------    ----------------               -----------------     ----------------
+        # TAILING             Read logs, Pause, Get status   Job complete          JOB_COMPLETE
+        #                                                    Else                  TAILING
+        # JOB_COMPLETE        Read logs, Pause               Any                   COMPLETE
+        # COMPLETE            Read logs, Exit                                      N/A
+        #
+        # Notes:
+        # - The JOB_COMPLETE state forces us to do an extra pause and read any items that
+        # got to Cloudwatch after the job was marked complete.
+        last_describe_job_call = time.time()
+        last_description = description
+
+        while True:
+            time.sleep(check_interval)
+
+            sec = sec + check_interval
+            if max_ingestion_time and sec > max_ingestion_time:
+                # ensure that the job gets killed if the max ingestion time is exceeded
+                raise AirflowException('SageMaker job took more than '
+                                       '%s seconds', max_ingestion_time)
```

Comment:
Same comment here about position of max time check

Reply:
Updated.